src/cockpit/channels/filesystem.py

```diff
@@ -268,3 +287,226 @@ def do_close(self):
         self._watch.close()
         self._watch = None
         self.close()
+
+
+class FsInfoChannel(Channel):
+    payload = 'fsinfo'
+
+    # Options (all get set in `do_open()`)
+    path: str
+    attrs: 'set[str]'
+    fnmatch: str
+    follow: bool
+    watch: bool
+
+    # State
+    current_value: JsonDict
+    effective_fnmatch: str = ''
+    fd: 'Handle | None' = None
+    pending: 'set[str] | None' = None
+    path_watch: 'PathWatch | None' = None
+    getattrs: 'Callable[[int, str], JsonDocument]'
+
+    @staticmethod
+    def make_getattrs(attrs: Sequence[str]) -> 'Callable[[int, str], JsonDocument | None]':
+        # Cached for the duration of the closure we're creating
+        @functools.lru_cache()
+        def get_user(uid: int) -> 'str | int':
+            try:
+                return pwd.getpwuid(uid).pw_name
+            except KeyError:
+                return uid
+
+        @functools.lru_cache()
+        def get_group(gid: int) -> 'str | int':
+            try:
+                return grp.getgrgid(gid).gr_name
+            except KeyError:
+                return gid
+
+        stat_types = {stat.S_IFREG: 'reg', stat.S_IFDIR: 'dir', stat.S_IFLNK: 'lnk', stat.S_IFCHR: 'chr',
+                      stat.S_IFBLK: 'blk', stat.S_IFIFO: 'fifo', stat.S_IFSOCK: 'sock'}
+        available_stat_getters = {
+            'type': lambda buf: buf and stat_types.get(stat.S_IFMT(buf.st_mode)),
+            'tag': tag_from_stat,
+            'mode': lambda buf: stat.S_IMODE(buf.st_mode),
+            'size': lambda buf: buf.st_size,
+            'uid': lambda buf: buf.st_uid,
+            'gid': lambda buf: buf.st_gid,
+            'mtime': lambda buf: buf.st_mtime,
+            'user': lambda buf: get_user(buf.st_uid),
+            'group': lambda buf: get_group(buf.st_gid),
+        }
+        stat_getters = tuple((key, available_stat_getters.get(key, lambda _: None)) for key in attrs)
+
+        def get_attrs(fd: int, name: str) -> 'JsonDict | None':
+            try:
+                buf = os.lstat(name, dir_fd=fd) if name else os.fstat(fd)
+            except FileNotFoundError:
+                return None
+            except OSError:
+                return {name: None for name, func in stat_getters}
```

Comment:
The top-level "error" is only if we fail to open the item referred to by the `path` option on the channel.  Failing to stat a single item in the directory is reported by returning `{}` for that item.

A side note about `{}` vs a dictionary full of `None`: you correctly identify that this is a patch which deletes all of the keys.  To be more explicit: the way the code is structured, the `getattr()` function is expected to return the full list of attributes every time.  That means that if there are attributes which we can't report, we need to return them as `None` to make sure they get removed from the existing value.

Should indeed be `_func` or `_`.

Reply:
I thought `self.getattrs()` would be used for `path` itself as well. If these errors are handled somewhere else, then returning `{}` (effectively) makes sense, but then I'd say we should log these as warnings -- once we opened the directory fd, it should be really exceptional that fstat() or lstat fails, as all of these entries are in the directory, not related to the file inode, no?

Reply:
`self.getattrs()` is used for the file itself, but only after the `O_PATH` fd has been successfully opened.  We call `fstat()`, which has no realistic chance to fail. Failures here are really about the subitems.

Note: we call this function on *any* change for the subitems, including deletion.  That's why we handle `FileNotFoundError` in a special way compared to all the other errors.

Other failures on statting items in the directory are indeed likely to be uncommon, but there's one possible case: a directory which has read permission but not search.  In that case, we'll see the directory entries, but won't be able to say anything about them.  That will result in `{}` being returned for each item, but at least the items will be listed.

It occurs to me that I didn't think about what should happen if search permission gets removed on a directory as we're watching it.  That's tricky.  Should we replace all of the existing files with `{}`?

And what if read permission gets removed?  Will inotify automatically shut down our watch in that case?

This probably could use a bit of investigation...

For the record, `ls` output in this case is the one where you see:

```
⬢[lis@toolbox cockpit]$ chmod a-x main; ls -l main
ls: cannot access 'main/containers': Permission denied
ls: cannot access 'main/doc': Permission denied
ls: cannot access 'main/examples': Permission denied
...
total 0
-????????? ? ? ? ?            ? aclocal.m4
-????????? ? ? ? ?            ? AUTHORS
-????????? ? ? ? ?            ? autogen.sh
d????????? ? ? ? ?            ? autom4te.cache
l????????? ? ? ? ?            ? bots
...
```

ie: it lists the files, but you can't see anything about them, relatively consistent with our result.

One difference between us and `ls` is that ls successfully identifies the file type still — that's surely a result of it using the `d_type` field of the dirent struct.  We could do that if we used `os.scandir()` instead of `os.listdir()`, but it would only work on the first iteration, which is why we favour just using `stat()`.

Reply:
I think I have a plan for what we could do here — effective treat the change in read/search permissions on the directory that we're monitoring as a identity-change-level type of event and rescan everything from new.

Detecting the change in the case that we start out with the read permission is not difficult.  Detecting the other direction is a bit of a different story, though, as inotify will prevent us from adding a watch to begin with if we can't read.

In any case: I think this needs to (and certainly can) be punted to a follow-up